### ISSUE TITLE
[ufo] Compile TrueType hinting to bytecode

### ISSUE THREAD

**Issue by jenskutilek** [LucasFonts/vfbLib#6]
Probably out of scope, but it would be nice to get production-ready UFOs

**Comment by typedev** [comment]
Hello Jens, any news or plans about this feature?

**Comment by jenskutilek** [comment]
No, not really. If you are adventurous, my [notes on TrueType hinting in FontLab 5](https://github.com/jenskutilek/fakelab/blob/main/doc/truetype-hinting/index.md) should contain everything that's needed.

**Comment by typedev** [comment]
Jens, thanks for the link, I hadn't seen it before. You've done a great job!
I started looking into it because I've worked a bit with the YGT hint editor, you might have heard of it. But unfortunately, it seems like all the editors use their own language and it all feels like a little Babel. And since the old vfb base is still significant, I'd like to find a way to transfer hints to UFO and edit them with independent tools, and the research you've done will be very helpful. I'll study your article, thanks!

**Comment by jenskutilek** [comment]
If you can open the VFB in FL5, you can get at the FL-generated TrueType code now, as of release v0.11.6.

Say you want to add TrueType hinting to UFO1.ufo, which is converted from a TT VFB with `vfb3ufo`.

Export a hinted TTF from FL5, import it again in FL5 and save as VFB. If you then convert this VFB to a UFO, let's call it UFO2.ufo, the TrueType instructions will be in the font and glyph lib. You can copy them from UFO2 to your original UFO1 (Look for `public.truetype.instructions` in the font and glyph libs, `openTypeGaspRangeRecords` and `openTypeHeadFlags` in the font info).

If you then compile a TTF from UFO1 with fontmake, it should have the TrueType instructions as long as you don't change the outline direction and don't attempt to remove overlaps:

```sh
fontmake -u UFO1.ufo -o ttf --keep-direction --keep-overlaps
```

**Comment by typedev** [comment]
Jens, hi! Thanks! I already solved this problem by creating my own instruction-to-assembler compiler based on your description -- a pure Python-based library. It was a great challenge, but without your help with the documentation, it wouldn't have been possible, so thank you so much!

**Comment by jenskutilek** [comment]
That's great to hear!
